pkg/common-controller/snapshot_controller.go

```diff
@@ -1655,3 +1666,27 @@ func (ctrl *csiSnapshotCommonController) checkAndSetInvalidSnapshotLabel(snapsho
 
 	return updatedSnapshot, nil
 }
+
+func (ctrl *csiSnapshotCommonController) getManagedByNode(pv *v1.PersistentVolume) (string, error) {
+	if pv.Spec.NodeAffinity == nil {
+		klog.V(5).Infof("NodeAffinity not set for pv %s", pv.Name)
+		return "", nil
+	}
+	nodeSelectorTerms := pv.Spec.NodeAffinity.Required
+
+	nodes, err := ctrl.client.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
```

Comment:
I think it's no needed to list all nodes when creating volumesnapshotcontent everytimes.

Reply:
We not only want to check whether that node exists or not, but more importantly,  we also need to find which node matches the affinity criteria. That's why we need to get the list of nodes to figure out which nodes match the criteria.

Reply:
@nearora-msft Can we do something similar to what is done in external-provisioner?
https://github.com/kubernetes-csi/external-provisioner/pull/524/files#diff-963c2f5e2076b500e70c47c13f135cdbb29b81f270bc76d1c99f823fecd1b510R271

Reply:
@xing-yang , I'm not sure I completely understand your suggestion. One difference that I can see between csi-provisioner and snapshot-controller is that csi-provisioner is deployed on each node, so it can use the environment variable NODE_NAME to get the node details, while in this implementation, the snapshot-controller will run as a single central instance and doesn't have any information about NODE_NAME. 

Let me known in case you meant something else.

Reply:
If there's only a single instance of it, why does it need to do something different for distributed snapshotting? But I don't know the snapshotter code, so I might be missing something entirely.

Reply:
So I am glad I am not the only one that got this wrong the first time, IMO @nearora-msft is right that we need to list all the nodes here. The controller code here is creating the VolumeSnapshotContent object, and then labeling it so that the side cars (csi driver or snapshot side car) can easily identify if they are the node needing to do something with it. 

In order to label the VolumeSnapshotContent with the correct node name, we look at the node affinity of the PV that is the source of the VolumeSnapshotContent. And match the affinity term there against nodes, if one of the nodes matches, we know that is the node we are looking for. 

Since each CSI driver can use its own term naming, we cant figure out ahead of time what that is, so we have to look for the node name, and if that matches we know we have the right term. So we have to fetch all the nodes, and compare until we find one that matches.

Reply:
@nearora-msft Can this be a node informer that is only created if distributed snapshotting is supported?

Reply:
@xing-yang , getManagedByNode will only be invoked if distributed snapshotting is enabled. Is there a particular reason that we should create an informer instead of making the List call?

Reply:
@nearora-msft Informer should be used for better performance since this is invoked for every create snapshot call when distributed snapshotting is enabled.

Reply:
Added an informer for listing nodes.